### PR TITLE
Start Solving For Type Variables (Locally)

### PR DESCRIPTION
**Pull request by CodaFi**
Adds two new forms of shorthand for closures.  This necessitated modeling type variables and starting the most rudimentary of solvers in the TypeChecker.

<strike>YOUR PARSER MAKES TYPE VARIABLES, I HOPE YOU'RE HAPPY</strike>

### CODE REVIEW

File: Sources/Parse/FunctionParser.swift [hunk 1]
```diff
@@ -148,6 +148,31 @@ extension Parser {
...
   }
+
+  func parseShorthandSignature() throws -> [ParamDecl] {
+    var args = [ParamDecl]()
+    while true {
+      let startLoc = sourceLoc
+      let name = try parseIdentifier()
+
+      let tv = DataType.freshTypeVariable
+      guard case let .typeVariable(tyName) = tv else {
```
Comment (harlanhaskins) [inline Sources/Parse/FunctionParser.swift]: This seems unnecessary, no?

**Comment by CodaFi** [comment]
Hm, [so this removes the ability to "coerce"](https://travis-ci.org/trill-lang/trill/builds/209583989#L1295).

**Comment by segiddins** [comment]
Can you also add an example to demonstrate what this will allow?

File: Sources/Base/Statements.swift [hunk 2]
```diff
@@ -75,6 +75,15 @@ class VarAssignDecl: Decl {
     superAttrs["mutable"] = mutable
     return superAttrs
   }
+
+  override var type: DataType {
+    set {
+      super.type = newValue
+    }
+    get {
+      return self.rhs?.type ?? super.type
```
Comment (segiddins) [inline Sources/Base/Statements.swift]: I think this previously was done in `Sema`?

**Comment by segiddins** [comment]
Pardon my ignorance, but what's the benefit of propagating the canonical types directly?

File: Sources/Base/Statements.swift (hunk 2 above)
Comment (CodaFi) [reply Sources/Base/Statements.swift]: "Done" in the sense of "an assign expression's type was being 'deduced' by Sema to enable this upcast, but should actually be a function of the right hand side with a coercion inserted by the typechecker".

**Comment by CodaFi** [comment]
Propagating sugar down the context is better for debugging, but means every access to the type needs to be guarded by a canonicalize call.  Really, it's a modeling problem.  We should be able to retain annotated types and all their sugar, solve constraint systems with canonical types, then apply and get rid of the propagation call.

**Comment by harlanhaskins** [comment]
I actually think we should remove that. It makes type checker errors worse by stripping the sugar in diagnostics.

**Comment by harlanhaskins** [comment]
I eventually want to be able to say:

```
error: cannot convert value of type 'Foo' (aka 'Array<(Int, String)>') to type 'Bar' (aka String)
```

**Comment by CodaFi** [comment]
Starting work on the solver.  We're going OutsideIn light (no GADTs, no existentials, no generalization).

File: Sources/Semantic Analysis/Solver.swift [hunk 3]
```diff
@@ -0,0 +1,133 @@
+//
+//  CSGen.swift
+//  Trill
+//
+//  Created by Robert Widmann on 3/14/17.
+//  Copyright © 2017 Harlan. All rights reserved.
+//
+
+final class Solver {
```
Comment (segiddins) [inline Sources/Semantic Analysis/Solver.swift:9]: `TypeSolver`?

File: Sources/Semantic Analysis/Solver.swift [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      }
+    }
+  }
+
+  final class Generator: ASTTransformer {
+    var goal: DataType? = nil
+    var env: [Identifier:DataType] = [:]
+    var constraints: [Constraint] = []
+
+    func byBinding(_ n : Identifier, _ t : DataType, _ f : () -> ()) {
```
Comment (segiddins) [inline Sources/Semantic Analysis/Solver.swift]: any chance you could use more descriptive variable names?

File: Sources/Semantic Analysis/Solver.swift [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        self.constraints.append(.Eq(goalType, self.goal!))
+      }
+      // let <ident> = <expr>
+      else if let e = expr.rhs {
+        // Generate 
+        goalType = DataType.freshMetaVariable
+        byBinding(expr.name, goalType, {
+          visit(e)
+        })
+        let phi = Solver.solveSystem(self.constraints)
```
Comment (segiddins) [inline Sources/Semantic Analysis/Solver.swift]: `phi` is unused

File: Sources/Semantic Analysis/Solver.swift [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      }
+      // let <ident> : <Type>
+      else if let t = expr.type {
+        // Take the type binding as fact and move on.
+        goalType = t
+        self.env[expr.name] = goalType
+      }
+      else {
+        fatalError()
```
Comment (segiddins) [inline Sources/Semantic Analysis/Solver.swift]: add a message to the error

File: Sources/Base/Types.swift [hunk 7]
```diff
@@ -166,20 +192,20 @@ func ==(lhs: DataType, rhs: DataType) -> Bool {
 }
 
 class Decl: ASTNode {
-  var type: DataType
+  var type: DataType?
```
Comment (segiddins) [inline Sources/Base/Types.swift]: might it make more sense to introduce an `error` or `unknown` type instead of making this optional?

File: trill.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -1584,6 +1587,7 @@
 				DC8EBB3A1E5FCFD500AE577C /* Mangler.swift in Sources */,
 				DC8EBB401E5FCFD500AE577C /* TypeChecker.swift in Sources */,
 				DC8EBAFE1E5FCFB200AE577C /* RuntimeIRGen.swift in Sources */,
+				82D6E13A1E78E6CA00C791DE /* Solver.swift in Sources */,
```
Comment (segiddins) [inline trill.xcodeproj/project.pbxproj:1698]: `TypeSolver.swift` ?

**Comment by CodaFi** [comment]
@segiddins I appreciate the review, but nothing here works yet!

**Comment by segiddins** [comment]
Understood, just trying to understand it 😅

File: Sources/Base/Types.swift (hunk 7 above)
Comment (CodaFi) [reply Sources/Base/Types.swift]: @harlanhaskins Either way I expect queries before typechecking to blow up and queries after to not have to see error types in the AST.  It might even be worth it to create a second fully-typed AST that is suitable for lowering and scale the original back to just TypeExprs.

File: Sources/Base/Types.swift [hunk 9]
```diff
@@ -36,7 +42,15 @@ enum DataType: CustomStringConvertible, Hashable {
     return .array(field: field, length: nil)
   }
   static let string = DataType.custom(name: "String")
-  
+  static var freshTypeVariable : DataType {
+    defer { DataType.typeVariablePool += 1 }
+    return .typeVariable(name: "T\(DataType.typeVariablePool)")
+  }
+  static var freshMetaVariable : DataType {
```
Comment (segiddins) [inline Sources/Base/Types.swift:49]: This seems like duplication?

File: stdlib/Mirror.tr [hunk 10]
```diff
@@ -72,7 +72,7 @@ type Mirror {
     }
   }
 
-  func print(to file: *FILE) {
+  func printToFile(_ file: *FILE) {
```
Comment (segiddins) [inline stdlib/Mirror.tr:75]: Why this change?

Comment (harlanhaskins) [reply stdlib/Mirror.tr:75]: Oops. I was debugging some interactions with overload resolution.

This PR isn't nearly complete. it's gonna be a much larger investment.

File: trill.xcodeproj/project.pbxproj (hunk 8 above)
Comment (harlanhaskins) [reply trill.xcodeproj/project.pbxproj:1698]: Yeah probably

**Review by harlanhaskins — COMMENTED** [review]
I'm going to close this PR and move the development onto the main repo.

File: Sources/Base/Types.swift (hunk 9 above)
Comment (CodaFi) [reply Sources/Base/Types.swift:49]: Meta variables are spawned and solved during unification, type variables are spawned and solved by CSGen.  You can merge both definitions, but I kept them separate because of the paper I was reading while writing this.
